File: backend/src/main/java/com/weedrice/whiteboard/domain/notification/web/NotificationSseEvents.java
package com.weedrice.whiteboard.domain.notification.web;

import java.util.Set;

/**
 * 알림 SSE 스트림이 내보내는 이벤트 이름의 단일 출처.
 *
 * <p>이름은 프론트엔드 {@code notificationStreamController.ts}의 분기와 1:1로 대응하는
 * 계약이다. REST 응답과 달리 SSE는 envelope 검증 장치가 없으므로, 여기 상수를 거치지 않은
 * 이름을 쓰면 {@code NotificationSseEventContractTest}가 실패한다.
 *
 * <p>이름을 추가·변경할 때는 프론트엔드 union 타입도 함께 고쳐야 한다.
 *
 * <p>테스트는 이 상수를 참조하지 말고 기대하는 wire 값을 리터럴로 적는다. 테스트가 상수를
 * 따라가면 값을 잘못 고쳐도 함께 움직여 회귀를 놓친다.
 */
public final class NotificationSseEvents {

    /** 구독 직후 연결 식별자를 전달한다. */
    public static final String CONNECT = "connect";

    /** 새 알림 한 건. */
    public static final String NOTIFICATION = "notification";

    /** 구독 중인 게시글의 댓글 변경. */
    public static final String COMMENT = "comment";

    /** 스페이스 단위로 댓글 구독을 무효화한다. */
    public static final String COMMENT_TOPIC_INVALIDATED = "comment-topic-invalidated";

    /** 댓글 구독 대상에 대한 접근 권한이 사라졌다. */
    public static final String COMMENT_TOPIC_ACCESS_REVOKED = "comment-topic-access-revoked";

    /** 관리자에 의해 상점 아이템의 판매 가능 여부가 변경됐다. */
    public static final String SHOP_ITEM_SALE_STATUS_CHANGED = "shop-item-sale-status-changed";

    public static final Set<String> ALL = Set.of(
            CONNECT,
            NOTIFICATION,
            COMMENT,
            COMMENT_TOPIC_INVALIDATED,
            COMMENT_TOPIC_ACCESS_REVOKED,
            SHOP_ITEM_SALE_STATUS_CHANGED);

    private NotificationSseEvents() {
    }
}

File: frontend/src/features/notifications/stream/notificationSseEvents.ts
/**
 * 알림 SSE 스트림의 이벤트 이름 계약.
 *
 * 값은 백엔드 `com.weedrice.whiteboard.domain.notification.web.NotificationSseEvents`와
 * 1:1로 대응하며 한쪽을 고치면 반드시 다른 쪽도 함께 고쳐야 한다.
 *
 * 가드는 타입이 아니라 테스트가 맡는다. SSE 이벤트 이름은 wire에서 임의 문자열로 도착하므로
 * `handleSseEvent`의 파라미터는 `string`일 수밖에 없고, 따라서 분기 누락이나 오타를
 * 타입 검사로 잡을 수 없다. 대신 `notificationSseEvents.spec.ts`가
 * (1) 백엔드 상수 집합과의 일치, (2) 모든 이름이 실제로 분기 처리되는지를 검증한다.
 */
export const NOTIFICATION_SSE_EVENTS = {
    /** 구독 직후 연결 식별자를 전달한다. */
    CONNECT: 'connect',
    /** 새 알림 한 건. */
    NOTIFICATION: 'notification',
    /** 구독 중인 게시글의 댓글 변경. */
    COMMENT: 'comment',
    /** 스페이스 단위 댓글 구독 무효화. */
    COMMENT_TOPIC_INVALIDATED: 'comment-topic-invalidated',
    /** 댓글 구독 대상 접근 권한 상실. */
    COMMENT_TOPIC_ACCESS_REVOKED: 'comment-topic-access-revoked',
    /** 관리자에 의한 상점 아이템 판매 상태 변경. */
    SHOP_ITEM_SALE_STATUS_CHANGED: 'shop-item-sale-status-changed',
} as const

/**
 * SSE 규격이 정한, `event:` 줄이 없는 프레임의 기본 이름.
 * 백엔드가 보내는 이름이 아니라 프로토콜 기본값이며 `notificationSseStream.ts`가 이 값을 채운다.
 * 백엔드가 이름 없는 data 프레임을 보내게 되더라도 알림으로 처리하기 위해 함께 받아 둔다.
 */
export const SSE_DEFAULT_EVENT_NAME = 'message'
